client/src/components/annotators/ImageAnnotator.vue

```diff
@@ -97,37 +114,104 @@ export default {
           this.syncedFrame = this.maxFrame;
           return;
         }
+        if (!this.imgs[this.frame].cached || this.loadingVideo) {
+          this.frame -= 1;
+          this.loadingVideo = true;
+          return;
+        }
         this.seek(this.frame);
         setTimeout(this.syncWithVideo, 1000 / this.frameRate);
       }
     },
-    cacheImage() {
+    async cacheImage() {
       const { frame } = this;
-      const max = Math.min(frame + 10, this.maxFrame);
       const { imgs } = this;
+      const cachedFrames = this.cacheSeconds * this.frameRate;
+      const min = Math.floor(Math.max(0, frame - cachedFrames * (1 - this.frontBackRatio)));
+      const max = Math.ceil(Math.min(frame + this.frontBackRatio * cachedFrames, this.maxFrame));
+      const frameDiff = Math.abs(this.frame - this.lastFrame);
+      const prevFrame = (this.frame < this.lastFrame);
       this.pendingImgs.forEach((imageAndFrame) => {
-        if (imageAndFrame[1] < frame || imageAndFrame[1] > max) {
+        // the current loading cache nees to be wiped out if we seek forward, backwards or
+        // if we are out of the current range of the cache
+        if (imageAndFrame[1] < min || imageAndFrame[1] > max || frameDiff > 2 || prevFrame) {
           imgs[imageAndFrame[1]] = null;
           // eslint-disable-next-line no-param-reassign
           imageAndFrame[0].src = '';
           this.pendingImgs.delete(imageAndFrame);
         }
       });
-      for (let i = frame; i <= max; i += 1) {
-        if (!imgs[i]) {
+      // if not playing we want the seeked to frame immediately loaded for UX
+      if (!this.playing && !imgs[frame] && frame > 0) {
+        await this.loadFrame(frame);
+      }
+      // Cache a new range based on the parameters
+      this.cacheNewRange(min, max);
+    },
+    /**
+     * Wraps loading of a single frame in a promise
+     * @param {int} frame number to be loaded
+     * @returns {Promise} resolves when the image from the frame is loaded
+     */
+    loadFrame(frame) {
+      return new Promise((resolve) => {
+        const img = new Image();
+        img.crossOrigin = 'Anonymous';
+        img.src = this.imageUrls[frame];
+        this.imgs[frame] = img;
```

Comment:
I haven't dealt with a lot of browser caching myself, so this may be a dumb question:

Why do you have to actually keep a reference to the returned image?  Isn't it enough to just make the request and then destroy the DOM element, and rely on the browser network cache to hit the next time the request is made?

Reply:
Cache might be a bad term for what we are actually doing.  We are loading the images so that they can be played back in sequence while avoiding async calls if possible.  In other instances that aren't as time dependent, I've loaded the images I know would be needed then discarded the image objects afterwards.
geoJS is expecting an image object each frame when it renders.  It is utilizing some of the stuff contained within a loaded image object such as `img.width` and `img.height` as well as `img.src`.  We could recreate these image objects before the frame is drawn but the assignment of an image.src **I think** would kick it to the event loop even if it is in the browser cache, so depending on framerate of the playback you may get black flickering (what I guess geoJS draws if the `Image.src` is invalid).